File: samples/vdsp_op/buildin_op_prof/buildin_op_prof.cpp

/*
 * Copyright (C) 2025 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <thread>

#include "common/cmdline.hpp"
#include "common/model_profiler.hpp"
#include "common/utils.hpp"
#include "opencv2/opencv.hpp"
#include "vaststreamx/vaststreamx.h"

class BuildInOperatorProf {
 public:
  explicit BuildInOperatorProf(const std::string& op_config,
                               uint32_t device_id = 0)
      : device_id_(device_id) {
    CHECK(vsx::SetDevice(device_id) == 0)
        << "SetDevice " << device_id << " failed";
    ops_ = vsx::Operator::LoadOpsFromJsonFile(op_config);
    CHECK(ops_.size() == 1)
        << "Only support 1 BuildIn Op. Now it's " << ops_.size() << std::endl;

    buildin_op_ = static_cast<vsx::BuildInOperator*>(ops_[0].get());
    auto attri_keys = buildin_op_->GetAttrKeys();
    // for (auto key : attri_keys) {
    //   std::cout << "key: " << key << std::endl;
    // }
    auto op_type = buildin_op_->GetOpType();
    // std::cout << "op type: " << op_type << std::endl;
    CHECK(op_type != vsx::BuildInOperatorType::kBERT_EMBEDDING_OP)
        << "Unsupport operator BERT_EMBEDDING_OP";

    buildin_op_->GetAttribute<vsx::AttrKey::kIimageWidth>(iimage_width_);
    buildin_op_->GetAttribute<vsx::AttrKey::kIimageWidthPitch>(
        iimage_width_pitch_);
    buildin_op_->GetAttribute<vsx::AttrKey::kIimageHeight>(iimage_height_);
    buildin_op_->GetAttribute<vsx::AttrKey::kIimageHeightPitch>(
        iimage_height_pitch_);
    // get image format
    if (vsx::HasAttribute(attri_keys, "kIimageFormat")) {
      vsx::BuildInOperatorAttrImageType format;
      buildin_op_->GetAttribute<vsx::AttrKey::kIimageFormat>(format);
      iimage_format_ = vsx::ConvertToVsxFormat(format);
    }
    if (vsx::HasAttribute(attri_keys, "kOimageFormat")) {
      vsx::BuildInOperatorAttrImageType format;
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageFormat>(format);
      oimage_format_ = vsx::ConvertToVsxFormat(format);
    } else {
      oimage_format_ = iimage_format_;
    }

    if (op_type == vsx::BuildInOperatorType::kSINGLE_OP_CVT_COLOR) {
      vsx::BuildInOperatorAttrColorCvtCode cvtcolor_code;
      buildin_op_->GetAttribute<vsx::AttrKey::kColorCvtCode>(cvtcolor_code);
      vsx::ConvertToVsxFormat(cvtcolor_code, iimage_format_, oimage_format_);
    }
    oimage_count_ = 1;
    if (op_type == vsx::BuildInOperatorType::kSINGLE_OP_BATCH_CROP_RESIZE) {
      buildin_op_->GetAttribute<vsx::AttrKey::kCropNum>(oimage_count_);
      int ow, owp, oh, ohp;
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidth>(ow);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidthPitch>(owp);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeight>(oh);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeightPitch>(ohp);

      for (int i = 0; i < oimage_count_; i++) {
        oimage_width_.push_back(ow);
        oimage_width_pitch_.push_back(owp);
        oimage_height_.push_back(oh);
        oimage_height_pitch_.push_back(ohp);
      }
    } else if (op_type == vsx::BuildInOperatorType::kSINGLE_OP_SCALE) {
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageCnt>(oimage_count_);
      for (int i = 0; i < oimage_count_; i++) {
        int ow, owp, oh, ohp;
        buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidth>(ow, i);
        buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidthPitch>(owp, i);
        buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeight>(oh, i);
        buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeightPitch>(ohp, i);
        oimage_width_.push_back(ow);
        oimage_width_pitch_.push_back(owp);
        oimage_height_.push_back(oh);
        oimage_height_pitch_.push_back(ohp);
      }
    } else if (op_type ==
               vsx::BuildInOperatorType::kSINGLE_OP_COPY_MAKE_BORDER) {
      int ow, oh;
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidth>(ow);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeight>(oh);
      oimage_width_.push_back(ow);
      oimage_height_.push_back(oh);
    } else if (op_type ==
               vsx::BuildInOperatorType::
                   kFUSION_OP_RGB_LETTERBOX_CVTCOLOR_NORM_TENSOR_EXT) {
      int left, right, bottom, top, width, height;
      buildin_op_->GetAttribute<vsx::AttrKey::kPaddingLeft>(left);
      buildin_op_->GetAttribute<vsx::AttrKey::kPaddingRight>(right);
      buildin_op_->GetAttribute<vsx::AttrKey::kPaddingTop>(top);
      buildin_op_->GetAttribute<vsx::AttrKey::kPaddingBottom>(bottom);
      buildin_op_->GetAttribute<vsx::AttrKey::kResizeWidth>(width);
      buildin_op_->GetAttribute<vsx::AttrKey::kResizeHeight>(height);

      oimage_width_.push_back(left + right + width);
      oimage_width_pitch_.push_back(left + right + width);
      oimage_height_.push_back(top + bottom + height);
      oimage_height_pitch_.push_back(top + bottom + height);
    } else if (op_type ==
               vsx::BuildInOperatorType::kFUSION_OP_RGB_CVTCOLOR_NORM_TENSOR) {
      oimage_width_.push_back(iimage_width_);
      oimage_height_.push_back(iimage_height_);
      iimage_format_ = vsx::BGR_PLANAR;
      oimage_format_ = vsx::RGB_PLANAR;
    } else if (op_type >= vsx::BuildInOperatorType::
                              kFUSION_OP_YUV_NV12_RESIZE_2RGB_NORM_TENSOR) {
      int ow, oh;
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidth>(ow);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeight>(oh);
      oimage_width_.push_back(ow);
      oimage_height_.push_back(oh);
    } else {
      int ow, owp, oh, ohp;
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidth>(ow);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageWidthPitch>(owp);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeight>(oh);
      buildin_op_->GetAttribute<vsx::AttrKey::kOimageHeightPitch>(ohp);

      oimage_width_.push_back(ow);
      oimage_width_pitch_.push_back(owp);
      oimage_height_.push_back(oh);
      oimage_height_pitch_.push_back(ohp);
    }

    if (op_type >=
        vsx::BuildInOperatorType::kFUSION_OP_YUV_NV12_RESIZE_2RGB_NORM_TENSOR) {
      odata_type_ = vsx::TypeFlag::kFloat16;
      oimage_format_ = vsx::ImageFormat::RGB_PLANAR;
    }

    // PrintParams();
  }
  void PrintParams() {
    std::cout << "iimage_width_:" << iimage_width_ << std::endl;
    std::cout << "iimage_width_pitch_:" << iimage_width_pitch_ << std::endl;
    std::cout << "iimage_height_:" << iimage_height_ << std::endl;
    std::cout << "iimage_height_pitch_:" << iimage_height_pitch_ << std::endl;
    std::cout << "oimage_count_:" << oimage_count_ << std::endl;

    std::cout << "oimage_width_: [ ";
    for (auto& num : oimage_width_) std::cout << num << " ";
    std::cout << "]\n";

    std::cout << "oimage_width_pitch_: [ ";
    for (auto& num : oimage_width_pitch_) std::cout << num << " ";
    std::cout << "]\n";

    std::cout << "oimage_height_: [ ";
    for (auto& num : oimage_height_) std::cout << num << " ";
    std::cout << "]\n";

    std::cout << "oimage_height_pitch_: [ ";
    for (auto& num : oimage_height_pitch_) std::cout << num << " ";
    std::cout << "]\n";

    std::cout << "iimage_format_:" << vsx::ImageFormatToString(iimage_format_)
              << std::endl;
    std::cout << "oimage_format_:" << vsx::ImageFormatToString(oimage_format_)
              << std::endl;

    std::cout << "odata_type_:" << odata_type_ << std::endl;
  }
  int Process(vsx::Image& image) {
    std::vector<vsx::Image> images = {image};
    return Process(images);
  }
  int Process(std::vector<vsx::Image>& images) { return ProcessImpl(images); }

  std::vector<vsx::Image> GetTestData(
      uint32_t bsize, uint32_t dtype, const vsx::Context& context,
      const std::vector<vsx::TShape>& input_shapes) {
    const auto& input_shape = input_shapes[0];
    std::vector<vsx::Image> images;
    images.reserve(bsize);
    int width, height;
    CHECK(input_shape.ndim() >= 2);
    height = input_shape[input_shape.ndim() - 2];
    width = input_shape[input_shape.ndim() - 1];
    auto image = vsx::Image(iimage_format_, width, height, context);
    for (uint32_t i = 0; i < bsize; i++) {
      images.push_back(image);
    }

    // make device memory image for output
    for (size_t i = 0; i < images.size(); i++) {
      for (int n = 0; n < oimage_count_; n++) {
        auto image_vacc =
            vsx::Image(oimage_format_, oimage_width_[n], oimage_height_[n],
                       vsx::Context::VACC(device_id_), 0, 0, odata_type_);
        output_images_vacc_.push_back(image_vacc);
      }
    }

    return images;
  }

 protected:
  virtual int ProcessImpl(const std::vector<vsx::Image>& images) {
    std::vector<vsx::Image> input_images_vacc;

    for (const auto& image : images) {
      if (image.GetContext().dev_type == vsx::Context::kCPU) {
        auto image_vacc =
            vsx::Image(image.Format(), image.Width(), image.Height(),
                       vsx::Context::VACC(device_id_));
        image_vacc.CopyFrom(image);
        input_images_vacc.push_back(image_vacc);
      } else {
        input_images_vacc.push_back(image);
      }
    }
    // // make device memory image for output
    // std::vector<vsx::Image> output_images_vacc;
    // for (size_t i = 0; i < images.size(); i++) {
    //   for (int n = 0; n < oimage_count_; n++) {
    //     auto image_vacc =
    //         vsx::Image(oimage_format_, oimage_width_[n], oimage_height_[n],
    //                    vsx::Context::VACC(device_id_), 0, 0, odata_type_);
    //     output_images_vacc.push_back(image_vacc);
    //   }
    // }

    // buildin operator execute
    buildin_op_->Execute({input_images_vacc}, output_images_vacc_);

    return 0;
  }

 public:
  int iimage_width_, iimage_width_pitch_, iimage_height_, iimage_height_pitch_;
  int oimage_count_ = 1;
  std::vector<int> oimage_width_, oimage_width_pitch_;
  std::vector<int> oimage_height_, oimage_height_pitch_;
  vsx::ImageFormat iimage_format_ = vsx::ImageFormat::YUV_NV12, oimage_format_;
  vsx::TypeFlag odata_type_ = vsx::TypeFlag::kUint8;

 protected:
  std::vector<std::shared_ptr<vsx::Operator>> ops_;
  vsx::BuildInOperator* buildin_op_ = nullptr;
  uint32_t device_id_;
  std::vector<vsx::Image> output_images_vacc_;
};

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("device_ids", 'd', "device id to run", false, "[0]");
  args.add<std::string>("op_config", '\0', "build in op config json", true);
  args.add<uint32_t>("instance", 'i',
                     "instance number or range for each device", false, 1);
  args.add<int>("iterations", '\0', "iterations count for one profiling", false,
                10240);
  args.add<std::string>("percentiles", '\0', "percentiles of latency", false,
                        "[50, 90, 95, 99]");
  args.add<bool>("input_host", '\0', "cache input data into host memory", false,
                 0);
  args.add<uint32_t>("batch_size", '\0', "batch_size", false, 1);
  args.parse_check(argc, argv);
  return args;
}

int main(int argc, char* argv[]) {
  auto args = ArgumentParser(argc, argv);
  auto device_ids = vsx::ParseVecUint(args.get<std::string>("device_ids"));
  auto op_config = args.get<std::string>("op_config");
  auto input_host = args.get<bool>("input_host");
  auto percentiles = vsx::ParseVecUint(args.get<std::string>("percentiles"));

  auto iterations = args.get<int>("iterations");
  auto instance = args.get<uint32_t>("instance");
  auto batch_size = args.get<uint32_t>("batch_size");

  uint32_t queue_size = 0;

  std::vector<std::shared_ptr<BuildInOperatorProf>> ops;
  ops.reserve(instance);
  std::vector<vsx::Context> contexts;
  for (uint32_t i = 0; i < instance; i++) {
    uint32_t device_id = device_ids[i % (device_ids.size())];
    if (input_host) {
      contexts.push_back(vsx::Context::CPU());
    } else {
      contexts.push_back(vsx::Context::VACC(device_id));
    }
    ops.push_back(std::make_shared<BuildInOperatorProf>(op_config, device_id));
  }
  int width = ops[0]->iimage_width_;
  int height = ops[0]->iimage_height_;
  vsx::TShape input_shape({3, height, width});
  vsx::ProfilerConfig config = {instance,      iterations,  batch_size,
                                vsx::kInt32,   device_ids,  contexts,
                                {input_shape}, percentiles, queue_size};
  vsx::ModelProfiler<BuildInOperatorProf> profiler(config, ops);
  std::cout << profiler.Profiling() << std::endl;

  return 0;
}

File: samples/vdsp_op/buildin_op_prof/buildin_op_prof.py
import vaststreamx as vsx
from typing import Union, List
import numpy as np
import os
import sys

attr = vsx.AttrKey

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../../..")
sys.path.append(common_path)

from common.model_profiler import ModelProfiler
from easydict import EasyDict as edict
import argparse
import ast
import common.utils as utils


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--op_config",
        default="",
        help="op config file",
    )
    parser.add_argument(
        "-d",
        "--device_ids",
        default="[0]",
        help="device ids to run",
    )
    parser.add_argument(
        "-i",
        "--instance",
        default=1,
        type=int,
        help="instance number for each device",
    )
    parser.add_argument(
        "--iterations",
        default=10240,
        type=int,
        help="iterations count for one profiling",
    )

    parser.add_argument(
        "--percentiles",
        default="[50, 90, 95, 99]",
        help="percentiles of latency",
    )
    parser.add_argument(
        "--input_host",
        default=0,
        type=int,
        help="cache input data into host memory",
    )
    args = parser.parse_args()
    return args


def cv_rgb_image_to_vastai(image_cv, device_id):
    assert len(image_cv.shape) >= 2
    h = image_cv.shape[0]
    w = image_cv.shape[1]
    if len(image_cv.shape) == 3:
        return vsx.create_image(
            image_cv, vsx.ImageFormat.BGR_INTERLEAVE, w, h, device_id
        )
    elif len(image_cv.shape) == 2:
        return vsx.create_image(image_cv, vsx.ImageFormat.GRAY, w, h, device_id)
    else:
        raise Exception("unsupported ndarray shape", image_cv.shape)


class BuildInOperatorProf:
    def __init__(self, op_config, device_id):
        self.device_id_ = device_id
        vsx.set_device(device_id)

        self.ops_ = vsx.Operator.load_ops_from_json_file(op_config)
        assert (
            len(self.ops_) == 1
        ), f"Only support 1 BuildIn Op. Now it's {len(self.ops_)}"
        self.buildin_op_ = self.ops_[0].cast_to_buildin_operator()

        assert (
            self.buildin_op_.op_type != vsx.OpType.BERT_EMBEDDING_OP
        ), "Unsupport operator BERT_EMBEDDING_OP"

        self.iimage_width_ = self.buildin_op_.get_attribute(attr.IIMAGE_WIDTH)
        self.iimage_height_ = self.buildin_op_.get_attribute(attr.IIMAGE_HEIGHT)
        self.iimage_width_pitch_ = self.buildin_op_.get_attribute(
            attr.IIMAGE_WIDTH_PITCH
        )
        self.iimage_height_pitch_ = self.buildin_op_.get_attribute(
            attr.IIMAGE_HEIGHT_PITCH
        )

        self.iimage_format_ = vsx.ImageFormat.YUV_NV12
        self.oimage_format_ = vsx.ImageFormat.YUV_NV12

        if "IIMAGE_FORMAT" in list(self.buildin_op_.attributes.keys()):
            imagetype = self.buildin_op_.get_attribute(attr.IIMAGE_FORMAT)
            self.iimage_format_ = utils.imagetype_to_vsxformat(imagetype)

        if "OIMAGE_FORMAT" in list(self.buildin_op_.attributes.keys()):
            imagetype = self.buildin_op_.get_attribute(attr.OIMAGE_FORMAT)
            self.oimage_format_ = utils.imagetype_to_vsxformat(imagetype)

        if self.buildin_op_.op_type == vsx.OpType.SINGLE_OP_CVT_COLOR:
            cvtcolor_code = self.buildin_op_.get_attribute(attr.COLOR_CVT_CODE)
            self.iimage_format_, self.oimage_format_, _ = (
                utils.cvtcolorcode_to_vsxformat(cvtcolor_code)
            )

        self.oimage_count_ = 1
        self.oimage_width_ = []
        self.oimage_height_ = []
        self.oimage_width_pitch_ = []
        self.oimage_height_pitch_ = []

        if self.buildin_op_.op_type == vsx.OpType.SINGLE_OP_BATCH_CROP_RESIZE:
            self.oimage_count_ = self.buildin_op_.get_attribute(attr.CROP_NUM)
            ow = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH)
            oh = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT)
            owp = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH_PITCH)
            ohp = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT_PITCH)
            for _ in range(self.oimage_count_):
                self.oimage_width_.append(ow)
                self.oimage_height_.append(oh)
                self.oimage_width_pitch_.append(owp)
                self.oimage_height_pitch_.append(ohp)
        elif self.buildin_op_.op_type == vsx.OpType.SINGLE_OP_SCALE:
            self.oimage_count_ = self.buildin_op_.get_attribute(attr.OIMAGE_CNT)
            for i in range(self.oimage_count_):
                ow = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH, i)
                oh = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT, i)
                owp = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH_PITCH, i)
                ohp = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT_PITCH, i)
                self.oimage_width_.append(ow)
                self.oimage_height_.append(oh)
                self.oimage_width_pitch_.append(owp)
                self.oimage_height_pitch_.append(ohp)
        elif self.buildin_op_.op_type == vsx.OpType.SINGLE_OP_COPY_MAKE_BORDER:
            ow = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH)
            oh = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT)
            self.oimage_width_.append(ow)
            self.oimage_height_.append(oh)
        elif (
            self.buildin_op_.op_type
            == vsx.OpType.FUSION_OP_RGB_LETTERBOX_CVTCOLOR_NORM_TENSOR_EXT
        ):
            left = self.buildin_op_.get_attribute(attr.PADDING_LEFT)
            right = self.buildin_op_.get_attribute(attr.PADDING_RIGHT)
            top = self.buildin_op_.get_attribute(attr.PADDING_TOP)
            bottom = self.buildin_op_.get_attribute(attr.PADDING_BOTTOM)
            width = self.buildin_op_.get_attribute(attr.RESIZE_WIDTH)
            height = self.buildin_op_.get_attribute(attr.RESIZE_HEIGHT)
            self.oimage_width_.append(left + right + width)
            self.oimage_height_.append(top + bottom + height)
            self.oimage_width_pitch_.append(left + right + width)
            self.oimage_height_pitch_.append(top + bottom + height)
        elif self.buildin_op_.op_type == vsx.OpType.FUSION_OP_RGB_CVTCOLOR_NORM_TENSOR:
            self.oimage_width_.append(self.iimage_width_)
            self.oimage_height_.append(self.iimage_height_)
            self.oimage_width_pitch_.append(self.iimage_width_)
            self.oimage_height_pitch_.append(self.iimage_height_)
            self.iimage_format_ = vsx.ImageFormat.BGR_PLANAR
            self.oimage_format_ = vsx.ImageFormat.RGB_PLANAR
        elif (
            self.buildin_op_.op_type
            >= vsx.OpType.FUSION_OP_YUV_NV12_RESIZE_2RGB_NORM_TENSOR
        ):
            ow = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH)
            oh = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT)
            self.oimage_width_.append(ow)
            self.oimage_height_.append(oh)
            self.oimage_width_pitch_.append(ow)
            self.oimage_height_pitch_.append(oh)
        else:
            ow = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH)
            oh = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT)
            owp = self.buildin_op_.get_attribute(attr.OIMAGE_WIDTH_PITCH)
            ohp = self.buildin_op_.get_attribute(attr.OIMAGE_HEIGHT_PITCH)
            self.oimage_width_.append(ow)
            self.oimage_height_.append(oh)
            self.oimage_width_pitch_.append(owp)
            self.oimage_height_pitch_.append(ohp)

        self.odata_type_ = vsx.TypeFlag.UINT8
        if (
            self.buildin_op_.op_type
            >= vsx.OpType.FUSION_OP_YUV_NV12_RESIZE_2RGB_NORM_TENSOR
        ):
            self.odata_type_ = vsx.TypeFlag.FLOAT16
            self.oimage_format_ = vsx.ImageFormat.RGB_PLANAR

        # self.PrintParams()

    def process(
        self, input: Union[List[np.ndarray], List[vsx.Image], np.ndarray, vsx.Image]
    ):
        if isinstance(input, list):
            if isinstance(input[0], np.ndarray):
                return self.process(
                    [cv_rgb_image_to_vastai(x, self.device_id_) for x in input]
                )
            else:
                return self.process_impl(input)
        return self.process([input])[0]

    def get_test_data(self, dtype, input_shape, batch_size, ctx="VACC"):
        assert len(input_shape) >= 2
        height = input_shape[-2]
        width = input_shape[-1]

        if self.iimage_format_ == vsx.ImageFormat.YUV_NV12:
            dummy_cv = np.random.randn(height * 3 // 2, width).astype(dtype)
        elif self.iimage_format_ == vsx.ImageFormat.RGB_INTERLEAVE:
            dummy_cv = np.random.randn(height, width * 3).astype(dtype)
        elif self.iimage_format_ == vsx.ImageFormat.BGR_INTERLEAVE:
            dummy_cv = np.random.randn(height, width * 3).astype(dtype)
        elif self.iimage_format_ == vsx.ImageFormat.GRAY:
            dummy_cv = np.random.randn(height, width).astype(dtype)
        else:
            dummy_cv = np.random.randn(height, width, 3).astype(dtype)

        if ctx == "CPU":
            return [dummy_cv] * batch_size
        else:

            vacc_dummy = vsx.create_image(
                dummy_cv, self.iimage_format_, width, height, self.device_id_
            )
            return [vacc_dummy] * batch_size

    def process_impl(self, inputs):
        scale = 1
        if self.odata_type_ == vsx.TypeFlag.FLOAT16:
            scale = 2
        output_info = [
            (self.oimage_format_, ow * scale, oh)
            for ow, oh in zip(self.oimage_width_, self.oimage_height_)
        ] * len(inputs)
        outs = self.buildin_op_.execute(
            images=inputs,
            output_info=output_info,
        )
        return outs

    def PrintParams(self):
        print(f"op_type:{self.buildin_op_.op_type}")
        print(f"self.attributes:{self.buildin_op_.attributes}")
        print(f"attribute keys:{list(self.buildin_op_.attributes.keys())}")
        print(f"self.iimage_width_:{self.iimage_width_}")
        print(f"self.iimage_height_:{self.iimage_height_}")
        print(f"self.iimage_width_pitch_:{self.iimage_width_pitch_}")
        print(f"self.iimage_height_pitch_:{self.iimage_height_pitch_}")
        print(f"self.oimage_width_:{self.oimage_width_}")
        print(f"self.oimage_height_:{self.oimage_height_}")
        print(f"self.oimage_width_pitch_:{self.oimage_width_pitch_}")
        print(f"self.oimage_height_pitch_:{self.oimage_height_pitch_}")
        print(f"self.oimage_count_:{self.oimage_count_}")
        print(f"self.iimage_format_:{self.iimage_format_}")
        print(f"self.oimage_format_:{self.oimage_format_}")
        print(f"self.odata_type_:{self.odata_type_}")


if __name__ == "__main__":
    args = argument_parser()
    op_config = args.op_config
    device_ids = ast.literal_eval(args.device_ids)
    instance = args.instance
    iterations = args.iterations
    queue_size = 0
    batch_size = 1
    input_host = args.input_host
    percentiles = ast.literal_eval(args.percentiles)

    ops = []
    contexts = []

    for i in range(instance):
        device_id = device_ids[i % len(device_ids)]
        op = BuildInOperatorProf(op_config=op_config, device_id=device_id)
        ops.append(op)
        if input_host:
            contexts.append("CPU")
        else:
            contexts.append("VACC")

    shape = (1, 3, ops[0].iimage_height_, ops[0].iimage_width_)
    config = edict(
        {
            "instance": instance,
            "iterations": iterations,
            "batch_size": batch_size,
            "data_type": "uint8",
            "device_ids": device_ids,
            "contexts": contexts,
            "input_shape": shape,
            "percentiles": percentiles,
            "queue_size": queue_size,
        }
    )
    profiler = ModelProfiler(config, ops)
    print(profiler.profiling())
